qiskit_nature/second_q/mappers/tapered_qubit_mapper.py

```diff
@@ -0,0 +1,197 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The Tapered Qubit Mapper."""
+
+from __future__ import annotations
+
+import logging
+from typing import cast, TYPE_CHECKING
+
+from qiskit.algorithms.list_or_dict import ListOrDict as ListOrDictType
+from qiskit.opflow import PauliSumOp
+from qiskit.quantum_info.analysis.z2_symmetries import Z2Symmetries
+from qiskit.quantum_info.operators import SparsePauliOp, Pauli
+
+from qiskit_nature.second_q.operators import SparseLabelOp
+
+from .qubit_mapper import QubitMapper, _ListOrDict
+
+if TYPE_CHECKING:
+    from qiskit_nature.second_q.problems import ElectronicStructureProblem
+
+logger = logging.getLogger(__name__)
+
+
+class TaperedQubitMapper(QubitMapper):  # pylint: disable=missing-class-docstring
+    """The wrapper around qubit mappers implementing the logic to reduce the size of a problem (operator)
+    based on mathematical `Z2Symmetries` that can be automatically detected in the operator.
+
+    The following attributes can be read and updated once the ``TaperedQubitMapper`` object has been
+    constructed.
+
+    Attributes:
+        mapper: Object defining the mapping of second quantized operators to Pauli operators.
+        z2symmetries: Symmetries to use to reduce the Pauli operators.
+    """
+
+    def __init__(
+        self,
+        mapper: QubitMapper,
+        z2symmetries: Z2Symmetries = Z2Symmetries([], [], []),
+    ):
+        """
+
+        Args:
+            mapper: ``QubitMapper`` object implementing the mapping of second quantized operators to
+                Pauli operators.
+            z2symmetries: ``Z2Symmetries`` object defining the symmetries to use to reduce the Pauli
+                operators.
+        """
+        super().__init__()
+        self.mapper: QubitMapper = mapper
+        self.z2symmetries = z2symmetries
+
+    @classmethod
+    def from_problem(
+        cls, mapper: QubitMapper, problem: "ElectronicStructureProblem"
+    ) -> "TaperedQubitMapper":
```

Comment:
I wonder if we can find a better name. It is just a little bit confusing to see this method name and then have a `QubitMapper` as the first argument..

Could also just be me to find it odd :thinking:

Reply:
When I look at the sample code in the main conversation it looks like two different ways to create this - the constructor, which takes a mapper and Z2Symmetries, and then this which takes a mapper and problem. I ended up wondering whether perhaps we could/should invert the relationship and have the problem support say a get_tapered_mapper(mapper) method. The logic makes use of calls across to problem, and by placing it there, should other problems be able to do this in the future they can add their own equivalents rather than having to somehow grow out some central location handling different problems. Maybe the build out of second_q_ops can then be managed/cached otherwise its building operators here, and mapping them just to find symmetries - the build out and mapping will then happen again in the main path when the mapper created here gets used right? Maybe there is some other integration which can do this more on the fly with the main operator build/out processing so we only do that once. Assuming I'm not missing something here.

Reply:
Yes this is indeed a good point. from_problem() basically uses the problem to obtain operators to obtain symmetries. The get_tapered_mapper(mapper) would reduce back and forth calls
I'll just mention a caveat I just thought of. For how it translates to code, I imagine something along the lines of
```
driver = PySCFDriver()
self.driver_result = driver.run()
mapper = ParityMapper(num_particles = self.driver_result.num_particles)
tapered_qubit_mapper = self.driver_result.get_tapered_mapper(mapper)
qubit_op = tapered_qubit_mapper.map(self.h2_op).primitive

```
I felt like - 1) Driver 2) Mapper 3) TaperedQubitMapper.from_problem() 4) TaperedQubitMapper.map() - was more linear than  - 1) Driver 2) Mapper 3) Driver.get_tapered_mapper() 4) TaperedQubitMapper.map() -

Reply:
Yeh basically this shifts the back and forth from inside the code to the workflow the user would have to deal with.
However, I still think that Steve has several good points here, especially considering the potential future integration with other problems.

Thus I would also vote for the `Problem.get_tapered_mapper(mapper)` design rather than the current approach. It does indeed feel more natural. This should also not be too surprising to users since the `symmetry_sector_locator` has always been on the problems (and this is effectively what enables them to support Z2Symmetry reductions). Thus, keeping this logic together on one object seems reasonable.